SecurityException: openInputStream() in FileManager#zipFiles()

Samsung Galaxy A71 Android 10 (SDK 29)

```ruby
java.lang.SecurityException: 
  at android.os.Parcel.createException (Parcel.java:2088)
  at android.os.Parcel.readException (Parcel.java:2056)
  at android.database.DatabaseUtils.readExceptionFromParcel (DatabaseUtils.java:188)
  at android.database.DatabaseUtils.readExceptionWithFileNotFoundExceptionFromParcel (DatabaseUtils.java:151)
  at android.content.ContentProviderProxy.openTypedAssetFile (ContentProviderNative.java:705)
  at android.content.ContentResolver.openTypedAssetFileDescriptor (ContentResolver.java:1694)
  at android.content.ContentResolver.openAssetFileDescriptor (ContentResolver.java:1510)
  at android.content.ContentResolver.openInputStream (ContentResolver.java:1194)
  at org.onionshare.android.files.FileManager.zipFiles (FileManager.kt:70)
  at org.onionshare.android.ShareManager.getSendPage (ShareManager.kt:118)
  at org.onionshare.android.ShareManager.access$getSendPage (ShareManager.kt:32)
  at org.onionshare.android.ShareManager$startSharing$2.invokeSuspend (ShareManager.kt:94)
  at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith (ContinuationImpl.kt:33)
  at kotlinx.coroutines.DispatchedTask.run (DispatchedTask.kt:106)
  at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely (CoroutineScheduler.kt:571)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask (CoroutineScheduler.kt:750)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker (CoroutineScheduler.kt:678)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run (CoroutineScheduler.kt:665)
```

Potentially related to https://code.briarproject.org/briar/briar/-/issues/2143 or https://code.briarproject.org/briar/briar/-/issues/2216

Samsung Galaxy J6+ Android 10 (SDK 29)

```ruby
java.lang.SecurityException: 
  at android.os.Parcel.createException (Parcel.java:2088)
  at android.os.Parcel.readException (Parcel.java:2056)
  at android.os.Parcel.readException (Parcel.java:2004)
  at android.app.IActivityManager$Stub$Proxy.getContentProvider (IActivityManager.java:5776)
  at android.app.ActivityThread.acquireProvider (ActivityThread.java:7123)
  at android.app.ContextImpl$ApplicationContentResolver.acquireUnstableProvider (ContextImpl.java:2893)
  at android.content.ContentResolver.acquireUnstableProvider (ContentResolver.java:2124)
  at android.content.ContentResolver.openTypedAssetFileDescriptor (ContentResolver.java:1678)
  at android.content.ContentResolver.openAssetFileDescriptor (ContentResolver.java:1510)
  at android.content.ContentResolver.openInputStream (ContentResolver.java:1194)
  at org.onionshare.android.files.FileManager.zipFiles (FileManager.kt:70)
  at org.onionshare.android.ShareManager.getSendPage (ShareManager.kt:118)
  at org.onionshare.android.ShareManager.access$getSendPage (ShareManager.kt:32)
  at org.onionshare.android.ShareManager$startSharing$2.invokeSuspend (ShareManager.kt:94)
  at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith (ContinuationImpl.kt:33)
  at kotlinx.coroutines.DispatchedTask.run (DispatchedTask.kt:106)
  at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely (CoroutineScheduler.kt:571)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask (CoroutineScheduler.kt:750)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker (CoroutineScheduler.kt:678)
  at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run (CoroutineScheduler.kt:665)
Caused by: android.os.RemoteException: 
  at com.android.server.am.ActivityManagerService.getContentProviderImpl (ActivityManagerService.java:8308)
  at com.android.server.am.ActivityManagerService.getContentProviderImpl (ActivityManagerService.java:8154)
  at com.android.server.am.ActivityManagerService.getContentProvider (ActivityManagerService.java:8784)
  at android.app.IActivityManager$Stub.onTransact (IActivityManager.java:2417)
  at com.android.server.am.ActivityManagerService.onTransact (ActivityManagerService.java:3359)
```

I can reproduce this on the Samsung Galaxy A10s using #21:

* Select a file, start sharing
* Back out of the activity - the activity and view model are destroyed but the foreground service is still running
* Relaunch the app from the launcher icon
* Stop sharing
* Start sharing again

```
2022-02-03 13:16:30.376 30833-30883/org.onionshare.android.nightly E/AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-3
    Process: org.onionshare.android.nightly, PID: 30833
    java.lang.SecurityException: Permission Denial: opening provider com.android.providers.media.MediaDocumentsProvider from ProcessRecord{4148957 30833:org.onionshare.android.nightly/u0a5} (pid=30833, uid=10005) requires that you obtain access using ACTION_OPEN_DOCUMENT or related APIs
        at android.os.Parcel.createExceptionOrNull(Parcel.java:2385)
        at android.os.Parcel.createException(Parcel.java:2369)
        at android.os.Parcel.readException(Parcel.java:2352)
        at android.os.Parcel.readException(Parcel.java:2294)
        at android.app.IActivityManager$Stub$Proxy.getContentProvider(IActivityManager.java:6683)
        at android.app.ActivityThread.acquireProvider(ActivityThread.java:7784)
        at android.app.ContextImpl$ApplicationContentResolver.acquireUnstableProvider(ContextImpl.java:3090)
        at android.content.ContentResolver.acquireUnstableProvider(ContentResolver.java:2491)
        at android.content.ContentResolver.openTypedAssetFileDescriptor(ContentResolver.java:1977)
        at android.content.ContentResolver.openAssetFileDescriptor(ContentResolver.java:1808)
        at android.content.ContentResolver.openInputStream(ContentResolver.java:1485)
        at org.onionshare.android.files.FileManager$zipFiles$2.invokeSuspend(FileManager.kt:71)
        at org.onionshare.android.files.FileManager$zipFiles$2.invoke(Unknown Source:8)
        at org.onionshare.android.files.FileManager$zipFiles$2.invoke(Unknown Source:4)
        at kotlinx.coroutines.flow.SafeFlow.collectSafely(Builders.kt:61)
        at kotlinx.coroutines.flow.AbstractFlow.collect(Flow.kt:212)
        at org.onionshare.android.ShareManager$startSharing$2.invokeSuspend(ShareManager.kt:209)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
     Caused by: android.os.RemoteException: Remote stack trace:
        at com.android.server.am.ActivityManagerService.getContentProviderImpl(ActivityManagerService.java:9027)
        at com.android.server.am.ActivityManagerService.getContentProviderImpl(ActivityManagerService.java:8866)
        at com.android.server.am.ActivityManagerService.getContentProvider(ActivityManagerService.java:9567)
        at android.app.IActivityManager$Stub.onTransact(IActivityManager.java:2843)
        at com.android.server.am.ActivityManagerService.onTransact(ActivityManagerService.java:3628)
```

I managed to reproduce this again while testing #27 on the Samsung A10s:
* Take a photo
* Open OnionShare, add the photo to the list of shared files
* Switch to the Gallery app, delete the photo ("move to recycle bin")
* Switch back to OnionShare, start sharing

```
2022-02-04 11:39:42.855 15932-16766/org.onionshare.android.nightly E/AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-1
    Process: org.onionshare.android.nightly, PID: 15932
    java.lang.SecurityException: Permission Denial: opening provider com.android.providers.media.MediaDocumentsProvider from ProcessRecord{b8c33ab 15932:org.onionshare.android.nightly/u0a5} (pid=15932, uid=10005) requires that you obtain access using ACTION_OPEN_DOCUMENT or related APIs
        at android.os.Parcel.createExceptionOrNull(Parcel.java:2385)
        at android.os.Parcel.createException(Parcel.java:2369)
        at android.os.Parcel.readException(Parcel.java:2352)
        at android.os.Parcel.readException(Parcel.java:2294)
        at android.app.IActivityManager$Stub$Proxy.getContentProvider(IActivityManager.java:6683)
        at android.app.ActivityThread.acquireProvider(ActivityThread.java:7784)
        at android.app.ContextImpl$ApplicationContentResolver.acquireUnstableProvider(ContextImpl.java:3090)
        at android.content.ContentResolver.acquireUnstableProvider(ContentResolver.java:2491)
        at android.content.ContentResolver.openTypedAssetFileDescriptor(ContentResolver.java:1977)
        at android.content.ContentResolver.openAssetFileDescriptor(ContentResolver.java:1808)
        at android.content.ContentResolver.openInputStream(ContentResolver.java:1485)
        at org.onionshare.android.files.FileManager$zipFiles$2.invokeSuspend(FileManager.kt:74)
        at org.onionshare.android.files.FileManager$zipFiles$2.invoke(Unknown Source:8)
        at org.onionshare.android.files.FileManager$zipFiles$2.invoke(Unknown Source:4)
        at kotlinx.coroutines.flow.SafeFlow.collectSafely(Builders.kt:61)
        at kotlinx.coroutines.flow.AbstractFlow.collect(Flow.kt:212)
        at org.onionshare.android.ShareManager$startSharing$2.invokeSuspend(ShareManager.kt:222)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
     Caused by: android.os.RemoteException: Remote stack trace:
        at com.android.server.am.ActivityManagerService.getContentProviderImpl(ActivityManagerService.java:9027)
        at com.android.server.am.ActivityManagerService.getContentProviderImpl(ActivityManagerService.java:8866)
        at com.android.server.am.ActivityManagerService.getContentProvider(ActivityManagerService.java:9567)
        at android.app.IActivityManager$Stub.onTransact(IActivityManager.java:2843)
        at com.android.server.am.ActivityManagerService.onTransact(ActivityManagerService.java:3628)
```

I'll try reproducing this on another phone to see if Samsung's "recycle bin" is relevant.

--- a/.gitignore
+++ b/.gitignore
@@ -14,3 +14,4 @@
 .cxx
 local.properties
 /Gemfile.lock
+/.bundle/config

--- a/Gemfile
+++ b/Gemfile
@@ -1,3 +1,6 @@
 source "https://rubygems.org"
 
 gem "fastlane"
+
+plugins_path = File.join(File.dirname(__FILE__), 'fastlane', 'Pluginfile')
+eval_gemfile(plugins_path) if File.exist?(plugins_path)

--- a/app/src/main/java/org/onionshare/android/ShareManager.kt
+++ b/app/src/main/java/org/onionshare/android/ShareManager.kt
@@ -10,6 +10,7 @@ import kotlinx.coroutines.GlobalScope
 import kotlinx.coroutines.Job
 import kotlinx.coroutines.cancel
 import kotlinx.coroutines.cancelAndJoin
+import kotlinx.coroutines.delay
 import kotlinx.coroutines.ensureActive
 import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.StateFlow
@@ -17,8 +18,10 @@ import kotlinx.coroutines.flow.collect
 import kotlinx.coroutines.flow.takeWhile
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.withContext
+import org.onionshare.android.files.FileErrorException
 import org.onionshare.android.files.FileManager
 import org.onionshare.android.files.FilesZipping
+import org.onionshare.android.files.totalSize
 import org.onionshare.android.server.SendFile
 import org.onionshare.android.server.SendPage
 import org.onionshare.android.server.WebserverManager
@@ -61,8 +64,7 @@ class ShareManager @Inject constructor(
 
         // not supporting selecting entire folders with sub-folders
         val filesAdded = fileManager.addFiles(uris, shareState.value.files)
-        val totalSize = filesAdded.files.sumOf { it.size }
-        _shareState.value = ShareUiState.FilesAdded(filesAdded.files, totalSize)
+        _shareState.value = ShareUiState.FilesAdded(filesAdded.files, filesAdded.files.totalSize)
     }
 
     fun removeFile(file: SendFile) {
@@ -73,8 +75,7 @@ class ShareManager @Inject constructor(
         if (newList.isEmpty()) {
             _shareState.value = ShareUiState.NoFiles
         } else {
-            val totalSize = newList.sumOf { it.size }
-            _shareState.value = ShareUiState.FilesAdded(newList, totalSize)
+            _shareState.value = ShareUiState.FilesAdded(newList)
         }
     }
 
@@ -96,6 +97,7 @@ class ShareManager @Inject constructor(
     }
 
     @OptIn(DelicateCoroutinesApi::class)
+    @Suppress("BlockingMethodInNonBlockingContext")
     private suspend fun startSharing() {
         if (startSharingJob?.isActive == true) {
             // TODO check if this always works as expected
@@ -143,7 +145,7 @@ class ShareManager @Inject constructor(
             } catch (e: IOException) {
                 LOG.warn("Error while startSharing ", e)
                 // launching stop on global scope to prevent deadlock when it waits for current job
-                GlobalScope.launch { stopSharing(error = true) }
+                GlobalScope.launch { stopSharing(exception = e) }
                 cancel("Error while startSharing", e)
             }
         }
@@ -179,7 +181,7 @@ class ShareManager @Inject constructor(
 
     private suspend fun stopSharing(
         complete: Boolean = false,
-        error: Boolean = false,
+        exception: Exception? = null,
     ) = withContext(Dispatchers.IO) {
         LOG.info("Stopping sharing...")
         if (startSharingJob?.isActive == true) {
@@ -194,10 +196,20 @@ class ShareManager @Inject constructor(
         val newState = when {
             files.isEmpty() -> ShareUiState.NoFiles
             complete -> ShareUiState.Complete(files, shareState.value.totalSize)
-            error -> ShareUiState.Error(files, shareState.value.totalSize)
+            exception is FileErrorException -> {
+                // remove errorFile from list of files, so user can try again
+                val newFiles = files.toMutableList().apply { remove(exception.file) }
+                ShareUiState.Error(newFiles, newFiles.totalSize, exception.file)
+            }
+            exception != null -> ShareUiState.Error(files, shareState.value.totalSize)
             else -> ShareUiState.FilesAdded(files, shareState.value.totalSize)
         }
         _shareState.value = newState
+        // special case handling for error state without file left
+        if (newState is ShareUiState.Error && newState.files.isEmpty()) {
+            delay(1000)
+            _shareState.value = ShareUiState.NoFiles
+        }
     }
 
     private fun SendFile.releaseUriPermission() {

--- a/app/src/main/java/org/onionshare/android/ShareManager.kt
+++ b/app/src/main/java/org/onionshare/android/ShareManager.kt
@@ -21,7 +21,6 @@ import kotlinx.coroutines.withContext
 import org.onionshare.android.files.FileErrorException
 import org.onionshare.android.files.FileManager
 import org.onionshare.android.files.FilesZipping
-import org.onionshare.android.files.totalSize
 import org.onionshare.android.server.SendFile
 import org.onionshare.android.server.SendPage
 import org.onionshare.android.server.WebserverManager
@@ -64,7 +63,7 @@ class ShareManager @Inject constructor(
 
         // not supporting selecting entire folders with sub-folders
         val filesAdded = fileManager.addFiles(uris, shareState.value.files)
-        _shareState.value = ShareUiState.FilesAdded(filesAdded.files, filesAdded.files.totalSize)
+        _shareState.value = ShareUiState.FilesAdded(filesAdded.files)
     }
 
     fun removeFile(file: SendFile) {
@@ -108,7 +107,7 @@ class ShareManager @Inject constructor(
             val files = shareState.value.files
             // call ensureActive() before any heavy work to ensure we don't continue when cancelled
             ensureActive()
-            _shareState.value = ShareUiState.Starting(files, shareState.value.totalSize, 0, 0)
+            _shareState.value = ShareUiState.Starting(files, 0, 0)
             try {
                 // TODO we might want to look into parallelizing what happens below (async {} ?)
                 // When the current scope gets cancelled, the async routine gets cancelled as well
@@ -116,7 +115,7 @@ class ShareManager @Inject constructor(
                 var sendPage: SendPage? = null
                 fileManager.zipFiles(files).collect { state ->
                     ensureActive()
-                    _shareState.value = ShareUiState.Starting(files, shareState.value.totalSize, state.progress, 0)
+                    _shareState.value = ShareUiState.Starting(files, state.progress, 0)
                     if (state.complete) sendPage = getSendPage(state)
                 }
                 val page = sendPage ?: error("SendPage was null")
@@ -127,15 +126,15 @@ class ShareManager @Inject constructor(
                 torManager.state.takeWhile { it !is TorState.Started }.collect { state ->
                     if (state is TorState.Starting) {
                         _shareState.value =
-                            ShareUiState.Starting(files, shareState.value.totalSize, 100, state.progress)
+                            ShareUiState.Starting(files, 100, state.progress)
                         onion = state.onion
                     }
                 }
-                _shareState.value = ShareUiState.Starting(files, shareState.value.totalSize, 100, 100)
+                _shareState.value = ShareUiState.Starting(files, 100, 100)
                 val onionAddress = onion ?: error("onion was null")
                 val url = "http://$onionAddress"
                 LOG.error("OnionShare URL: $url") // TODO remove before release
-                val sharing = ShareUiState.Sharing(files, shareState.value.totalSize, url)
+                val sharing = ShareUiState.Sharing(files, url)
                 // TODO properly manage tor and webserver state together
                 ensureActive()
                 // collecting from StateFlow will only return when coroutine gets cancelled
@@ -195,14 +194,14 @@ class ShareManager @Inject constructor(
         val files = shareState.value.files
         val newState = when {
             files.isEmpty() -> ShareUiState.NoFiles
-            complete -> ShareUiState.Complete(files, shareState.value.totalSize)
+            complete -> ShareUiState.Complete(files)
             exception is FileErrorException -> {
                 // remove errorFile from list of files, so user can try again
                 val newFiles = files.toMutableList().apply { remove(exception.file) }
-                ShareUiState.Error(newFiles, newFiles.totalSize, exception.file)
+                ShareUiState.Error(newFiles, exception.file)
             }
-            exception != null -> ShareUiState.Error(files, shareState.value.totalSize)
-            else -> ShareUiState.FilesAdded(files, shareState.value.totalSize)
+            exception != null -> ShareUiState.Error(files)
+            else -> ShareUiState.FilesAdded(files)
         }
         _shareState.value = newState
         // special case handling for error state without file left

--- a/app/src/main/java/org/onionshare/android/files/FileManager.kt
+++ b/app/src/main/java/org/onionshare/android/files/FileManager.kt
@@ -79,6 +79,9 @@ class FileManager @Inject constructor(
                         } catch (e: FileNotFoundException) {
                             LOG.warn("Error while opening file: ", e)
                             throw FileErrorException(file)
+                        } catch (e: SecurityException) {
+                            LOG.warn("Error while opening file: SecurityException")
+                            throw FileErrorException(file)
                         }
                     }
                 }

--- a/app/src/main/java/org/onionshare/android/files/FileManager.kt
+++ b/app/src/main/java/org/onionshare/android/files/FileManager.kt
@@ -17,6 +17,7 @@ import org.onionshare.android.R
 import org.onionshare.android.server.SendFile
 import org.slf4j.LoggerFactory.getLogger
 import java.io.File
+import java.io.FileNotFoundException
 import java.io.IOException
 import java.util.zip.ZipEntry
 import java.util.zip.ZipOutputStream
@@ -29,6 +30,8 @@ private val LOG = getLogger(FileManager::class.java)
 
 data class FilesAdded(val files: List<SendFile>)
 data class FilesZipping(val files: List<SendFile>, val zip: File, val progress: Int, val complete: Boolean = false)
+
+class FileErrorException(val file: SendFile) : IOException()
 
 @Singleton
 class FileManager @Inject constructor(
@@ -67,12 +70,16 @@ class FileManager @Inject constructor(
                         val progress = ((i + 1) / files.size.toFloat() * 100).roundToInt()
                         // TODO remove before release
                         LOG.debug("Zipping next file $progress/100: ${file.basename}")
-                        // TODO handle FileNotFoundException and tell user about problem / remove file
-                        ctx.contentResolver.openInputStream(file.uri)?.use { inputStream ->
-                            zipStream.putNextEntry(ZipEntry(file.basename))
-                            inputStream.copyTo(zipStream)
+                        try {
+                            ctx.contentResolver.openInputStream(file.uri)?.use { inputStream ->
+                                zipStream.putNextEntry(ZipEntry(file.basename))
+                                inputStream.copyTo(zipStream)
+                            }
+                            emit(FilesZipping(files, zipFile, progress))
+                        } catch (e: FileNotFoundException) {
+                            LOG.warn("Error while opening file: ", e)
+                            throw FileErrorException(file)
                         }
-                        emit(FilesZipping(files, zipFile, progress))
                     }
                 }
             }
@@ -89,3 +96,5 @@ class FileManager @Inject constructor(
     }
 
 }
+
+val List<SendFile>.totalSize get() = sumOf { it.size }

--- a/app/src/main/java/org/onionshare/android/ui/FileList.kt
+++ b/app/src/main/java/org/onionshare/android/ui/FileList.kt
@@ -183,7 +183,7 @@ fun FileListPreviewDark() {
                 SendFile("foo bar file", "1337 KiB", 1, Uri.parse("/foo"), null),
             )
             val mutableState = remember {
-                mutableStateOf(ShareUiState.FilesAdded(files, files.sumOf { it.size }))
+                mutableStateOf(ShareUiState.FilesAdded(files))
             }
             FileList(Modifier, mutableState, {}) {}
         }
@@ -199,9 +199,8 @@ fun FileListPreview() {
             SendFile("bar", "42 MiB", 2, Uri.parse("/bar"), "video/mp4"),
             SendFile("foo bar", "23 MiB", 3, Uri.parse("/foo/bar"), null),
         )
-        val totalSize = files.sumOf { it.size }
         val mutableState = remember {
-            mutableStateOf(ShareUiState.FilesAdded(files, totalSize))
+            mutableStateOf(ShareUiState.FilesAdded(files))
         }
         FileList(Modifier, mutableState, {}) {}
     }

--- a/app/src/main/java/org/onionshare/android/ui/MainUi.kt
+++ b/app/src/main/java/org/onionshare/android/ui/MainUi.kt
@@ -225,7 +225,7 @@ fun DefaultPreview() {
     )
     OnionshareTheme {
         MainUi(
-            stateFlow = MutableStateFlow(ShareUiState.FilesAdded(files, 1337L)),
+            stateFlow = MutableStateFlow(ShareUiState.FilesAdded(files)),
             onFabClicked = {},
             onFileRemove = {},
             onRemoveAll = {},

--- a/app/src/main/java/org/onionshare/android/ui/MainUi.kt
+++ b/app/src/main/java/org/onionshare/android/ui/MainUi.kt
@@ -88,9 +88,15 @@ fun MainUi(
             delay(750)
             scaffoldState.bottomSheetState.expand()
         }
-        if (state.value is ShareUiState.Error) {
-            val text = stringResource(R.string.share_error_snackbar_text)
-            val action = stringResource(R.string.share_error_snackbar_action)
+        val uiState = state.value
+        if (uiState is ShareUiState.Error) {
+            val errorFile = uiState.errorFile
+            val text = if (errorFile != null) {
+                stringResource(R.string.share_error_file_snackbar_text, errorFile.basename)
+            } else {
+                stringResource(R.string.share_error_snackbar_text)
+            }
+            val action = if (uiState.files.isEmpty()) null else stringResource(R.string.share_error_snackbar_action)
             LaunchedEffect("showSnackbar") {
                 val snackbarResult = scaffoldState.snackbarHostState.showSnackbar(
                     message = text,
@@ -100,22 +106,22 @@ fun MainUi(
                 if (snackbarResult == SnackbarResult.ActionPerformed) onSheetButtonClicked()
             }
         }
-        if (!state.value.collapsableSheet && scaffoldState.bottomSheetState.isCollapsed) {
+        if (!uiState.collapsableSheet && scaffoldState.bottomSheetState.isCollapsed) {
             // ensure the bottom sheet is visible
-            LaunchedEffect(state.value) {
+            LaunchedEffect(uiState) {
                 scaffoldState.bottomSheetState.expand()
             }
         }
         BottomSheetScaffold(
             topBar = { ActionBar(R.string.app_name) },
             floatingActionButton = {
-                Fab(state.value, scaffoldState.bottomSheetState, onFabClicked)
+                Fab(uiState, scaffoldState.bottomSheetState, onFabClicked)
             },
-            sheetGesturesEnabled = state.value.collapsableSheet,
+            sheetGesturesEnabled = uiState.collapsableSheet,
             sheetPeekHeight = bottomSheetPeekHeight,
             sheetShape = RoundedCornerShape(topStart = 16.dp, topEnd = 16.dp),
             scaffoldState = scaffoldState,
-            sheetContent = { BottomSheet(state.value, onSheetButtonClicked) }
+            sheetContent = { BottomSheet(uiState, onSheetButtonClicked) }
         ) {
             MainContent(stateFlow, offset, onFileRemove, onRemoveAll)
         }

--- a/app/src/main/java/org/onionshare/android/ui/ShareBottomSheet.kt
+++ b/app/src/main/java/org/onionshare/android/ui/ShareBottomSheet.kt
@@ -217,7 +217,7 @@ fun ShareBottomSheetReadyPreview() {
     OnionshareTheme {
         Surface(color = MaterialTheme.colors.background) {
             BottomSheet(
-                state = ShareUiState.FilesAdded(emptyList(), 0L),
+                state = ShareUiState.FilesAdded(emptyList()),
                 onSheetButtonClicked = {},
             )
         }
@@ -230,7 +230,7 @@ fun ShareBottomSheetStartingPreview() {
     OnionshareTheme {
         Surface(color = MaterialTheme.colors.background) {
             BottomSheet(
-                state = ShareUiState.Starting(emptyList(), 0L, 25, 50),
+                state = ShareUiState.Starting(emptyList(), 25, 50),
                 onSheetButtonClicked = {},
             )
         }
@@ -245,7 +245,6 @@ fun ShareBottomSheetSharingPreview() {
             BottomSheet(
                 state = ShareUiState.Sharing(
                     emptyList(),
-                    0L,
                     "http://openpravyvc6spbd4flzn4g2iqu4sxzsizbtb5aqec25t76dnoo5w7yd.onion/",
                 ),
                 onSheetButtonClicked = {},
@@ -266,7 +265,7 @@ fun ShareBottomSheetCompletePreview() {
     OnionshareTheme {
         Surface(color = MaterialTheme.colors.background) {
             BottomSheet(
-                state = ShareUiState.Complete(emptyList(), 0L),
+                state = ShareUiState.Complete(emptyList()),
                 onSheetButtonClicked = {},
             )
         }

--- a/app/src/main/java/org/onionshare/android/ui/ShareUiState.kt
+++ b/app/src/main/java/org/onionshare/android/ui/ShareUiState.kt
@@ -3,26 +3,23 @@ package org.onionshare.android.ui
 import org.onionshare.android.files.totalSize
 import org.onionshare.android.server.SendFile
 
-sealed class ShareUiState(open val files: List<SendFile>, open val totalSize: Long) {
+sealed class ShareUiState(val files: List<SendFile>) {
 
     open val allowsModifyingFiles = true
     open val collapsableSheet = false
+    val totalSize: Long = files.totalSize
 
-    object NoFiles : ShareUiState(emptyList(), 0L)
+    object NoFiles : ShareUiState(emptyList())
 
-    data class FilesAdded(
-        override val files: List<SendFile>,
-        override val totalSize: Long,
-    ) : ShareUiState(files, totalSize) {
-        constructor(files: List<SendFile>) : this(files, files.totalSize)
-    }
+    class FilesAdded(
+        files: List<SendFile>,
+    ) : ShareUiState(files)
 
-    data class Starting(
-        override val files: List<SendFile>,
-        override val totalSize: Long,
+    class Starting(
+        files: List<SendFile>,
         val zipPercent: Int,
         val torPercent: Int,
-    ) : ShareUiState(files, totalSize) {
+    ) : ShareUiState(files) {
         override val allowsModifyingFiles = false
         val totalProgress: Float
             get() {
@@ -36,24 +33,21 @@ sealed class ShareUiState(open val files: List<SendFile>, open val totalSize: Lo
         }
     }
 
-    data class Sharing(
-        override val files: List<SendFile>,
-        override val totalSize: Long,
+    class Sharing(
+        files: List<SendFile>,
         val onionAddress: String,
-    ) : ShareUiState(files, totalSize) {
+    ) : ShareUiState(files) {
         override val allowsModifyingFiles = false
         override val collapsableSheet = true
     }
 
-    data class Complete(
-        override val files: List<SendFile>,
-        override val totalSize: Long,
-    ) : ShareUiState(files, totalSize)
+    class Complete(
+        files: List<SendFile>,
+    ) : ShareUiState(files)
 
-    data class Error(
-        override val files: List<SendFile>,
-        override val totalSize: Long,
+    class Error(
+        files: List<SendFile>,
         val errorFile: SendFile? = null,
-    ) : ShareUiState(files, totalSize)
+    ) : ShareUiState(files)
 
 }

--- a/app/src/main/java/org/onionshare/android/ui/ShareUiState.kt
+++ b/app/src/main/java/org/onionshare/android/ui/ShareUiState.kt
@@ -1,5 +1,6 @@
 package org.onionshare.android.ui
 
+import org.onionshare.android.files.totalSize
 import org.onionshare.android.server.SendFile
 
 sealed class ShareUiState(open val files: List<SendFile>, open val totalSize: Long) {
@@ -12,7 +13,9 @@ sealed class ShareUiState(open val files: List<SendFile>, open val totalSize: Lo
     data class FilesAdded(
         override val files: List<SendFile>,
         override val totalSize: Long,
-    ) : ShareUiState(files, totalSize)
+    ) : ShareUiState(files, totalSize) {
+        constructor(files: List<SendFile>) : this(files, files.totalSize)
+    }
 
     data class Starting(
         override val files: List<SendFile>,
@@ -50,6 +53,7 @@ sealed class ShareUiState(open val files: List<SendFile>, open val totalSize: Lo
     data class Error(
         override val files: List<SendFile>,
         override val totalSize: Long,
+        val errorFile: SendFile? = null,
     ) : ShareUiState(files, totalSize)
 
 }

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -28,5 +28,6 @@
     <string name="sharing_channel_name">File sharing notification</string>
     <string name="sharing_notification_title">Currently sharing files</string>
     <string name="share_error_snackbar_text">Could not start OnionShare</string>
+    <string name="share_error_file_snackbar_text">Could not open %s. Removed.</string>
     <string name="share_error_snackbar_action">Retry</string>
 </resources>

--- a/fastlane/Fastfile
+++ b/fastlane/Fastfile
@@ -50,7 +50,7 @@ platform :android do
 
 	# Github: Ensure that tag exists and push it
 	git_tag_exists(tag: version_name)
-	push_git_tags(tag: version_name)
+	push_git_tags(remote: "upstream", tag: version_name)
 
     upload_to_play_store(
         track: 'beta',

